pkg/config/viper.go

```diff
@@ -16,243 +16,267 @@ import (
 func (c *safeConfig) GetBool(key string) bool {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetBool(key)
 }
 
-// GetInt is wrapped for concurrent access
+// GetInt wraps Viper for concurrent access
 func (c *safeConfig) GetInt(key string) int {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetInt(key)
 }
 
-// GetInt64 is wrapped for concurrent access
+// GetInt64 wraps Viper for concurrent access
 func (c *safeConfig) GetInt64(key string) int64 {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetInt64(key)
 }
 
-// GetFloat64 is wrapped for concurrent access
+// GetFloat64 wraps Viper for concurrent access
 func (c *safeConfig) GetFloat64(key string) float64 {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetFloat64(key)
 }
 
-// GetTime is wrapped for concurrent access
+// GetTime wraps Viper for concurrent access
 func (c *safeConfig) GetTime(key string) time.Time {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetTime(key)
 }
 
-// GetDuration is wrapped for concurrent access
+// GetDuration wraps Viper for concurrent access
 func (c *safeConfig) GetDuration(key string) time.Duration {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetDuration(key)
 }
 
-// GetStringSlice is wrapped for concurrent access
+// GetStringSlice wraps Viper for concurrent access
 func (c *safeConfig) GetStringSlice(key string) []string {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetStringSlice(key)
 }
 
-// GetStringMap is wrapped for concurrent access
+// GetStringMap wraps Viper for concurrent access
 func (c *safeConfig) GetStringMap(key string) map[string]interface{} {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetStringMap(key)
 }
 
-// GetStringMapString is wrapped for concurrent access
+// GetStringMapString wraps Viper for concurrent access
 func (c *safeConfig) GetStringMapString(key string) map[string]string {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetStringMapString(key)
 }
 
-// GetStringMapStringSlice is wrapped for concurrent access
+// GetStringMapStringSlice wraps Viper for concurrent access
 func (c *safeConfig) GetStringMapStringSlice(key string) map[string][]string {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetStringMapStringSlice(key)
 }
 
-// GetSizeInBytes is wrapped for concurrent access
+// GetSizeInBytes wraps Viper for concurrent access
 func (c *safeConfig) GetSizeInBytes(key string) uint {
 	c.RLock()
 	defer c.RUnlock()
 	return c.Viper.GetSizeInBytes(key)
 }
 
-// SetEnvPrefix is wrapped for concurrent access
+// SetEnvPrefix wraps Viper for concurrent access, and keeps the envPrefix for
+// future reference
 func (c *safeConfig) SetEnvPrefix(in string) {
 	c.Lock()
 	defer c.Unlock()
 	c.Viper.SetEnvPrefix(in)
+	c.envPrefix = in
 }
 
-// BindEnv is wrapped for concurrent access
+// BindEnv wraps Viper for concurrent access, and adds tracking of the configurable env vars
 func (c *safeConfig) BindEnv(input ...string) error {
 	c.RLock()
 	defer c.RUnlock()
+	if len(input) == 1 {
```

Comment:
Why do it only when input length is 1 ?

Reply:
this logic that tracks which env vars are configurable pre-existed, I just moved it around. It only works when `BindEnv` is used with one argument (in this case `BindEnv` deduces the env var key from the name of the config option, for example it adds a `DD_` prefix, and our logic later makes the same assumptions)

Reply:
Could point me to it ? I can't find where we pulled this from (should we remove it from where it is ?) ?

I also think that since we now have a wrapper around viper we should update this logic with to handle every arguments. I'm afraid we will forget about it and it will come back as a bug later.

Reply:
this logic was added here: https://github.com/DataDog/datadog-agent/pull/2294 (originally in `config.go`, I removed it from there)

I'll move the related tests to `viper_test.go` and add a FIXME

Reply:
done, resolving this